Add Apache Commons StringUtils refaster template replacements

AlekSimpson opened the pull request:
## What's changed?
Refaster rule recipes to replace Apache Commons StringUtils methods with JDK internals where possible.

## What's your motivation?
Fixes #267

## Anyone you would like to review specifically?
@timtebeek 

### Checklist
- [x] I've added unit tests to cover both positive and negative cases
- [x] I've added the license header to any new files through `./gradlew licenseFormat`
- [x] I've used the IntelliJ auto-formatter on affected files
- [ ] I've updated the documentation (if applicable)

knutwannheden commented on src/main/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtils.java:
```diff
@@ -0,0 +1,357 @@
+    public static class Equals {
+        @BeforeTemplate
+        boolean before(String s) {
+            return StringUtils.equals(s, "string");
```
Here you should change the signature so that you have `(String s1, String s2)` and then use those. Otherwise the recipe will only ever do anything if you actually call the method with the `"string"` literal you specify here. I think this same comment also applies to some of your other Refaster templates.

Unfortunately our Refaster support doesn't yet allow checking for the reverse order.

knutwannheden commented on src/main/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtils.java:
```diff
@@ -0,0 +1,357 @@
+
+        @AfterTemplate
+        boolean after(String s) {
+            return s.equals("string");
```
Can you verify that this is corrects with respect to nullability? Otherwise I suggest to use `Objects.equal()` here.

knutwannheden commented on src/main/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtils.java:
```diff
@@ -0,0 +1,357 @@
+    public static class Left {
+        @BeforeTemplate
+        String before(String s) {
+            return StringUtils.left(s, 5);
```
Also here for `5` instead introduce a second `int l` param.

knutwannheden commented on src/main/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtils.java:
```diff
@@ -0,0 +1,357 @@
+
+        @AfterTemplate
+        String after(String s) {
+            return s.substring(0, 5);
```
Haven't checked if we need to deal with `null` here. Should be checked for all templates, I think.

knutwannheden reviewed (commented):
I have added a first set of review comments.

timtebeek commented on src/main/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtils.java:
```diff
@@ -0,0 +1,361 @@
+
+        @AfterTemplate
+        boolean after(String s) {
+            return s.isEmpty();
```
I think this requires a null check too
```suggestion
            return s ==null || s.isEmpty();
```

AlekSimpson commented:
Ok the latest commit has all the changes made based off of the feedback given.

AlekSimpson commented:
I am still seeing errors when I run the tests. It tells me that `an enclosing instance that contains org.openrewrite.java.migrate.apache.commons.lang.ApacheCommonsStringUtilsRecipes.DefaultStringRecipe` is required. I'm not to sure what that error is saying to be honest.

knutwannheden commented:
> I am still seeing errors when I run the tests. It tells me that `an enclosing instance that contains org.openrewrite.java.migrate.apache.commons.lang.ApacheCommonsStringUtilsRecipes.DefaultStringRecipe` is required. I'm not to sure what that error is saying to be honest.

You probably need to declare the Refaster template classes as static.

AlekSimpson commented:
I have made them all static. I think the issue could be because I have the a local version of `rewrite-templating` to get some fixes earlier so I could make some progress while I waited for the fixes to make it to the main branch.

knutwannheden commented:
Meanwhile quite a few fixes have been integrated. I suggest you try with the latest integration build.

timtebeek commented on src/main/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtils.java:
```diff
@@ -0,0 +1,365 @@
+    }
+
+    @SuppressWarnings("ConstantValue")
+    public static class IsEmpty {
+        @BeforeTemplate
+        boolean before(String s) {
+            return StringUtils.isEmpty(s);
+        }
+
+        @AfterTemplate
+        boolean after(String s) {
+            return s == null || s.isEmpty();
+        }
+    }
+
+    public static class Replace {
```
Let's sort all nested classes here, such that it's easier to find them. Your IDEA likely has a sort-members option.

AlekSimpson replied:
Where can I find that option?

timtebeek replied:
Can get it to work here; the option appears to be named something like rearrange, but you might need to look at how to get that working, or just quickly drag and drop from the side bar.

timtebeek commented on src/main/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtils.java:
```diff
@@ -0,0 +1,365 @@
+        }
+    }
+
+    // TODO: These two methods don't generate the recipe templates right
+    //public static class LeftPad {
+    //    @BeforeTemplate
+    //    String before(String s, int l) {
+    //        return StringUtils.leftPad(s, l);
+    //    }
+
+    //    @AfterTemplate
+    //    String after(String s, int l) {
+    //        return String.format("%" + l + "s", s);
+    //    }
+    //}
+
+    //public static class RightPad {
+    //    @BeforeTemplate
+    //    String before(String s, int l) {
+    //        return StringUtils.rightPad(s, l);
+    //    }
+
+    //    @AfterTemplate
+    //    String after(String s, int l) {
+    //        return String.format("%" + (-l) + "s", s);
+    //    }
+    //}
```
I think we can try these again after the recent fixes.

AlekSimpson replied:
Seems like there still is an issue with these recipes. The error I get is `error: unclosed string literal .builder("String.format(\"%\" #{any(int)} l + \"s#{any(java.lang.String)}", s)");`

AlekSimpson replied:
Seems like the last '"' after `String)}` should also be escaped

AlekSimpson replied:
Well it happens every time I regenerate the recipes so I assume its something in rewrite-templating. This is on my desktop which didn't ever have the local rewrite-templating dependency so I don't think it could be an issue with that. If you uncomment the code and try to rebuild do you see the error?

timtebeek replied:
Confirmed; we can ignore this one for now then, and focus on the others to get a first batch through.

timtebeek commented on src/test/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtilsTest.java:
```diff
@@ -23,58 +24,273 @@
+    }
+
+    @Test
+    @Disabled("Argument variant not covered yet")
```
I think the corresponding recipe was missing here.

knutwannheden commented:
I suggest we also try to declare them as private, so that there won't be quite as many user-visible tiny recipes in the catalog.

timtebeek commented:
Given the amount of repetition in the tests I propose we instead adopt a pattern where we test multiple recipes in one test:
```java
@Test
void uberTest() {
    rewriteRun(
      //language=java
      java(
        """
          import org.apache.commons.lang3.StringUtils;
          
          class Foo {
              String in = "foo";
              String defaultString = StringUtils.defaultString(in);
              boolean isEmpty = StringUtils.isEmpty(in);
              String[] split = StringUtils.split(in);
              //String[] split2 = StringUtils.split(in, ", ");
              boolean equals = StringUtils.equals(in, "string");
          }
          """,
        """
          import java.util.Objects;
                        
          class Foo {
              String in = "foo";
              String defaultString = Objects.toString(in);
              boolean isEmpty = in == null || in.isEmpty();
              String[] split = in.split(" ");
              //String[] split2 = StringUtils.split(in, ", ");
              boolean equals = Objects.equals(in, "string");
          }
          """
      )
    );
}
```
Given the low complexity I think that's good-enough for these types of recipes & tests, and helps reduce the amount of code we produce and maintain.

AlekSimpson commented:
> I suggest we also try to declare them as private, so that there won't be quite as many user-visible tiny recipes in the catalog.

Do you mean the refaster class templates?

AlekSimpson commented:
> Given the amount of repetition in the tests I propose we instead adopt a pattern where we test multiple recipes in one test:
> 
> ```java
> @Test
> void uberTest() {
>     rewriteRun(
>       //language=java
>       java(
>         """
>           import org.apache.commons.lang3.StringUtils;
>           
>           class Foo {
>               String in = "foo";
>               String defaultString = StringUtils.defaultString(in);
>               boolean isEmpty = StringUtils.isEmpty(in);
>               String[] split = StringUtils.split(in);
>               //String[] split2 = StringUtils.split(in, ", ");
>               boolean equals = StringUtils.equals(in, "string");
>           }
>           """,
>         """
>           import java.util.Objects;
>                         
>           class Foo {
>               String in = "foo";
>               String defaultString = Objects.toString(in);
>               boolean isEmpty = in == null || in.isEmpty();
>               String[] split = in.split(" ");
>               //String[] split2 = StringUtils.split(in, ", ");
>               boolean equals = Objects.equals(in, "string");
>           }
>           """
>       )
>     );
> }
> ```
> 
> Given the low complexity I think that's good-enough for these types of recipes & tests, and helps reduce the amount of code we produce and maintain.

Do we want to group them together in groups of 4 or 5?

knutwannheden commented:
> > I suggest we also try to declare them as private, so that there won't be quite as many user-visible tiny recipes in the catalog.
> 
> Do you mean the refaster class templates?

Yes. The nested Refaster template classes. But for testing purposes package private would be slightly better. I don't really know which visibility will cause a recipe to get listed in the catalog.

timtebeek commented:
> Do we want to group them together in groups of 4 or 5? Or were you imagining that they would all for the most part be in one test?

We can start with all in one; and if there's an interesting variants you want to test for more complicated recipes then I'd cover those in a separate test.

timtebeek commented on src/test/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtilsTest.java:
```diff
@@ -45,6 +45,20 @@ void defaultString() {
+                  String[] out = StringUtils.split(in);
+                  String[] out = StringUtils.split(in, ", ");
+                  boolean out = StringUtils.equals(in, "string");
+                  String out = StringUtils.chop(in);
+                  String out = StringUtils.replace(in, "o", "z");
+                  String out = StringUtils.strip(in);
+                  String out = StringUtils.join(in);
+                  String out = StringUtils.deleteWhitespace(in);
+                  String out = StringUtils.abbreviate(in, 5);
+                  String out = StringUtils.trimToEmpty(in);
+                  String out = StringUtils.substringAfter(in, ",");
+                  String out = StringUtils.right(in, 5);
+                  String out = StringUtils.mid(in, 2, 4);
```
This won't compile with the left side variables all named the same, right? Perhaps best to reassign per type, such that you have multiple statements `str = StringUtils.xyz(in, ...)` and `arr = StringUtils.split(in);`

AlekSimpson replied:
Hm ya, I can fix this. Its odd though because I do remember being able to run the test and it passed fine. I wonder why this wasn't a bigger issue?

timtebeek replied:
I think we don't validate that we can actually compile the output of the recipe, because that's not always the case when you migrate but need a different version of a dependency on the classpath suddenly. Hence why we need to make sure the code looks at least plausible, ideally compiles with a manual inspection.

AlekSimpson commented:
Ok I have finished adding all the refaster template classes for all `StringUtils` methods that could be replaced. Most of them pass, a couple don't work due to a bug with `rewrite-templating` and the `Chomp` recipe does not work either. `Chomp` could possibly be not working due to another `rewrite-templating` bug. 

I also found the class arrangement tabs in IDEA to sort the refaster template classes. I'm not sure what criteria to use to sort the class though. @timtebeek how did you want the classes to be sorted?

timtebeek commented:
Figured best to sort alphabetically, that way it's easiest to spot any that are missing.

knutwannheden commented on src/main/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtils.java:
```diff
@@ -0,0 +1,607 @@
+
+        @AfterTemplate
+        String after(String s, int width) {
+            return s.substring(0, Math.min(s.length(), width));
```
I believe there are many overloads of `abbreviate()`. This particular one adds a trailing ellipsis. So you probably want something like:
```suggestion
            return s == null || s.length() <= width ? s : s.substring(0, width - 3) + "...";
```

knutwannheden commented on src/main/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtils.java:
```diff
@@ -0,0 +1,607 @@
+
+        @AfterTemplate
+        String after(String s) {
+            return s.substring(0, 1).toUpperCase() + s.substring(1);
```
I believe this is wrong with respect to nullability. I believe most of `StringUtils` methods accept `null` as input, so you should probably in most cases use the ternary operator to guard against that.

knutwannheden commented on src/main/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtils.java:
```diff
@@ -0,0 +1,607 @@
+        }
+    }
+
+    private static class CountMatches {
```
I think I would probably remove this rule. While it probably is possible to do something like this, we would have to properly escape the pattern (as the input to `countMatches()` isn't a regex) and also using regex has a much higher computational overhead.

We can however provide a rule for `countMatches(String, char)` by using something like `s.chars().filter(c -> c == pattern).count()` (or `s.codePoints().filter(c -> c == pattern).count()`).

knutwannheden commented on src/main/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtils.java:
```diff
@@ -0,0 +1,607 @@
+
+        @AfterTemplate
+        boolean after(String s, String suffix) {
+            return s.regionMatches(true, 0, suffix, 0, suffix.length());
```
This doesn't yet account for nulls, but should otherwise be correct:
```suggestion
            return s.regionMatches(true, s.length() - suffix.length(), suffix, 0, suffix.length());
```

knutwannheden commented on src/main/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtils.java:
```diff
@@ -0,0 +1,607 @@
+        int after(String s, String search) {
+            return IntStream.range(0, s.length())
+                    .filter(i -> search.indexOf(s.charAt(i)) >= 0)
+                    .findFirst()
```
```suggestion
                    .min()
```

knutwannheden commented on src/main/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtils.java:
```diff
@@ -0,0 +1,607 @@
+        }
+
+        boolean after(String s) {
+            return s.matches("^[a-zA-Z0-9\\s]*$");
```
I suspect this isn't correct with respect to Unicode characters (i.e. non ASCII characters). I suggest to instead use a `Stream` based on `String#chars()` like you do in `IsAlpha`.

knutwannheden reviewed (commented):
Didn't check all rules, but added a few comments.

timtebeek commented on build.gradle.kts:
```diff
@@ -51,6 +52,7 @@ dependencies {
     testImplementation("commons-codec:commons-codec:1.+")
     testImplementation("org.apache.commons:commons-lang3:3.+")
```
For dependencies pulled up to `implementation` we don't need to declare them again for tests.
```suggestion
```

timtebeek commented on src/test/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtilsTest.java:
```diff
@@ -0,0 +1,188 @@
+              class Foo {
+                  void test() {
+                      String in = "foo";
+                      String outString = StringUtils.defaultString(in);
+                      boolean outBool = StringUtils.isEmpty(in);
+                      String[] outArray = StringUtils.split(in);
+                      outArray = StringUtils.split(in, ", ");
+                      outBool = StringUtils.equals(in, "string");
+                      outString = StringUtils.chop(in);
+                      outString = StringUtils.replace(in, "o", "z");
+                      outString = StringUtils.strip(in);
+                      outString = StringUtils.join(in);
+                      outString = StringUtils.deleteWhitespace(in);
+                      outString = StringUtils.abbreviate(in, 5);
+                      outString = StringUtils.trimToEmpty(in);
+                      outString = StringUtils.substringAfter(in, ",");
+                      outString = StringUtils.right(in, 5);
+                      outString = StringUtils.mid(in, 2, 4);
+                      outString = StringUtils.stripStart(in, "chars");
+                      outString = StringUtils.uncapitalize(in);
+                      outString = StringUtils.capitalize(in);
+                      outString = StringUtils.removeEnd(in, "remove");
+                      int outInt = StringUtils.countMatches(in, "pattern");
+                      outString = StringUtils.trimToNull(in);
+                      outInt = StringUtils.indexOfAny(in, "search");
+                      outString = StringUtils.swapCase(in);
+                      outBool = StringUtils.isAlpha(in);
+                      outBool = StringUtils.isAlphaSpace(in);
```
I'd sort these based on the method called as well, such that it's again easier to spot what might not be covered by tests yet.

AlekSimpson commented:
@timtebeek Ok, I've commented out a bunch of the templates that I am not 100% sure are accurate replacements. I also added a test to explore the question of whether replacements should be surrounded in parentheses. 

I also think having all the tests grouped together into one test is a little tough to manage. Currently when I run the super test it throws a Java error but its not easy to tell which recipe(s) is/are causing it. Maybe its worth it to keep them in groups of 20? This its easier to track down the source of errors if they arise. 

I also wonder if it might be useful for large scale refaster recipe development to include an annotation that can disable specific templates. Maybe something like `@DisableTemplate`? This way refaster templates wouldn't have to be commented out like in this PR they could just be disable until ready. Its just a thought I had today that I thought was worth mentioning.

timtebeek commented on src/test/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtilsTest.java:
```diff
@@ -82,31 +97,46 @@ void test() {
+                      bool = in.equalsIgnoreCase("other");
+                      bool = Objects.equals(in, "other");
+                      str = Objects.equals(in, "other");
+                      bool = in == null ? false : in.matches("^[a-zA-Z0-9\\s]*$");
+                      bool = in == null ? false : in.chars().allMatch(Character::isAlphabetic);
+                      bool = in == null ? false : in.matches("[a-zA-Z\\s]+");
+                      bool = in == null ? false : in.chars().allMatch(Character::isLetter);
```
On these four my IDE has some suggestions for simplification; might be an Ultimate feature; are you using that as a student?
![image](https://github.com/openrewrite/rewrite-migrate-java/assets/1027334/6c839b2b-9b11-4e4d-93f6-d3840df9522c)

AlekSimpson replied:
Oh no I'm using the community edition. I will look into downloading the ultimate edition today.

timtebeek commented on src/test/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtilsTest.java:
```diff
@@ -82,31 +97,46 @@ void test() {
+                      String str = in == null || in.length() <= 10 ? in : in.substring(0, 10 - 3) + "...";
+                      str = in == null ? null : in.substring(0, 1).toUpperCase() + in.substring(1);
+                      str = in == null ? null : in.substring(0, in.length() - 1);
+                      str = in == null || "search" == null ? null : in.contains("search");
```
This one doesn't compile.

timtebeek commented on src/test/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtilsTest.java:
```diff
@@ -82,31 +97,46 @@ void test() {
+                      str = in == null || "search" == null ? null : in.contains("search");
+                      str = (int) (in == null ? 0 : in.chars().filter(c -> c == "pattern").count());
+                      str = in == null ? null : in.replaceAll("\\s+", "");
+                      boolean bool = in == null ? null : in.replaceAll("\\s+", "");
```
This one doesn't compile.

timtebeek commented on src/test/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtilsTest.java:
```diff
@@ -82,31 +97,46 @@ void test() {
+                      str = in == null ? null : in.split(", ");
+                      str = in == null ? null : (in.endsWith("suffix") ? in.substring(0, in.lastIndexOf("suffix")) : in);
+                      str = in == null ? null : (in.startsWith("chars") ? in.substring("chars".length()) : in);
+                      str = in == null || "prefix" == null ? null : in.startsWith("prefix");
```
This one doesn't compile.

timtebeek commented on src/test/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtilsTest.java:
```diff
@@ -82,31 +97,46 @@ void test() {
+                      str = in == null ? null : in.substring(0, 3) + "overlay" + in.substring(3);
+                      str = in == null ? null : (in.endsWith("remove") ? in.substring(0, in.length() - "remove".length()) : in);
+                      str = in == null ? null : new String(new char[3]).replace("\\0", in);
+                      str = in == null ? null : in.replaceFirst(Pattern.quote("search"), "replacement");
```
Missing Pattern import

timtebeek commented on src/test/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtilsTest.java:
```diff
@@ -0,0 +1,266 @@
+            """
+              class Foo {
+                  void test(String s) {
+                      String test = (s == null ? null : in.trim()).toString();
```
@knutwannheden this test currently fails; it's similar to a case we've discussed before on whether the `JavaTemplate`
 should itself detect whether a statements needs to be wrapped in braces. In this synthetic example we're changing behavior: before it would have thrown a NPE when the input is `null`, after non-braced replacement it no longer would. We should not make such behavioral changes.
```suggestion
                      String test = s == null ? null : in.trim().toString();
```
https://ge.openrewrite.org/s/jwi2f75wlzjts/tests/task/:test/details/org.openrewrite.java.migrate.apache.commons.lang.ApacheCommonsStringUtilsTest/canCallMethodOnResult()?top-execution=1

timtebeek commented on src/test/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtilsTest.java:
```diff
@@ -0,0 +1,266 @@
+            """
+              class Foo {
+                  void test(String s) {
+                      String test = !(s != null && !s.equalsIgnoreCase("other"));
```
Another variant @knutwannheden ; this one actually produces code that does not compile, but we should maintain the previous behavior.
```suggestion
                      String test = !(s != null && !s.equalsIgnoreCase("other"));
```
https://ge.openrewrite.org/s/jwi2f75wlzjts/tests/task/:test/details/org.openrewrite.java.migrate.apache.commons.lang.ApacheCommonsStringUtilsTest/invertedBooleanHandled()?top-execution=1

timtebeek replied:
Fixed in https://github.com/openrewrite/rewrite-templating/pull/22/

timtebeek commented:
We need a merge of this one to fix the parentheses cases:
- https://github.com/openrewrite/rewrite-templating/pull/22

timtebeek commented on src/test/java/org/openrewrite/java/migrate/apache/commons/lang/ApacheCommonsStringUtilsTest.java:
```diff
@@ -0,0 +1,287 @@
+                            
+              class Foo {
+                  void test(String s) {
+                      String test = StringUtils.strip(bar()).toString();
```
Logged as
- https://github.com/openrewrite/rewrite-templating/issues/27

timtebeek commented on build.gradle.kts:
```diff
@@ -49,8 +50,8 @@ dependencies {
     testImplementation("commons-codec:commons-codec:1.+")
-    testImplementation("org.apache.commons:commons-lang3:3.+")
     testImplementation("org.apache.maven.shared:maven-shared-utils:3.+")
+    testImplementation("org.codehaus.plexus:plexus-utils:3.+")
```
```suggestion
```
